api/saml/provider.py

```diff
@@ -0,0 +1,335 @@
+                        },
+                        {
+                            "language": "es",
+                            "height": 16,
+                            "width": 16,
+                            "value": "https://idp.saml.net/logo.png"
+                        }
+                    ],
+                }
+            ]
+        }
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: Authentication Flow object for use in an Authentication for OPDS document
+        :rtype: Dict
+        """
+        flow_doc = {
+            'type': self.FLOW_TYPE,
+            'description': self.NAME,
+            'links': []
+        }
+
+        configuration_serializer = SAMLConfigurationSerializer(self.external_integration(db))
+        metadata_serializer = SAMLMetadataSerializer(self.external_integration(db))
+        configuration = SAMLConfiguration(configuration_serializer, metadata_serializer)
+
+        for identity_provider in configuration.identity_providers:
+            link = {
+                'rel': 'authenticate',
+                'href': self._create_authenticate_url(db, identity_provider.entity_id),
+                'display_names': self._join_ui_info_items(identity_provider.ui_info.display_names),
+                'descriptions': self._join_ui_info_items(identity_provider.ui_info.descriptions),
+                'information_urls': self._join_ui_info_items(identity_provider.ui_info.information_urls),
+                'privacy_statement_urls': self._join_ui_info_items(
+                    identity_provider.ui_info.privacy_statement_urls),
+                'logo_urls': self._join_ui_info_items(identity_provider.ui_info.logo_urls)
+            }
+
+            flow_doc['links'].append(link)
+
+        return flow_doc
+
+    def _create_token(self, db, patron, token):
+        """Creates a Credential object that ties the given patron to the
+        given provider token.
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :param patron: Patron object
+        :type patron: Patron
+
+        :param token: Token containing SAML subject's metadata
+        :type token: Dict
+
+        :return: Credential object
+        :rtype: Credential
+        """
+        data_source, ignore = self._get_token_data_source(db)
+        duration = datetime.timedelta(days=self.token_expiration_days)
```

Comment:
Per our discussion above, this is where to put the TTL you get from the IdP, since (I believe) this token contains the actual SAML assertion you got from the IdP.

Reply:
Set credential.expiration to either AuthnStatement/@SessionNotOnOrAfter or SubjectConfirmationData/@NotOnOrAfter